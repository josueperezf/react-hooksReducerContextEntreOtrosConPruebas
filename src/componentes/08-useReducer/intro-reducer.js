const estadoInicial =[{
    id: 1,
    tarea : 'comprar pan',
    done: false
}];


const miFuncion = (state = estadoInicial, accion)=>{
    if(accion?.type === 'agregar') {
        return [...state, accion.payload]
    }
    return state;
}


let tareas = miFuncion();

const nuevaTarea = {
    id: 2,
    tarea : 'comprar leche',
    done: false
};

const agregarTareaAccion = {
    type:   'agregar',
    payload: nuevaTarea
}


tareas = miFuncion(tareas, agregarTareaAccion );

console.log(tareas);