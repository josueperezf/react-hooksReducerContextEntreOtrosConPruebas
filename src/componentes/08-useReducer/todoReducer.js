export const todoReducer = ( state = [], action )=>{
    switch (action.type) {
        case 'add':
            return [...state, action.payload];
        case 'delete':
            return state.filter((tarea)=>tarea.id !== action.payload);
        case 'toggle':
                return state.map((tarea)=> (tarea.id === action.payload)? {...tarea,done: !tarea.done} : tarea)
        /* case 'toggle':
            return state.map((tarea)=>{
                if(tarea.id === action.payload){
                    return {
                        ...tarea,
                        done: !tarea.done
                    } 
                }else {
                    return tarea
                }
            });*/
        default: 
            return state;
    }
}