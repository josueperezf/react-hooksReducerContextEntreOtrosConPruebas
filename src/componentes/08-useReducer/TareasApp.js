import React, { useEffect, useReducer } from 'react'
import './styles.css';
import { todoReducer } from './todoReducer';
import { TareaLista } from './componentes/TareaLista';
import { AgregarTarea } from './componentes/AgregarTarea';

const init = ()=>{
    return JSON.parse(localStorage.getItem('tareas')) || [];
}
export const TareasApp = () => {

    const [tareas, dispatch] = useReducer(todoReducer, [], init)
    
    

    // useEffect se esta colocando como un Listener que se dispara cuando cambia el valor de tareas
    useEffect(() => {
        localStorage.setItem('tareas', JSON.stringify(tareas));
    }, [tareas]);

    const handleAgregarTarea = (nuevaTarea)=>{
        dispatch({
            type:'add',
            payload: nuevaTarea
        });
    }

    const handleDelete = (tareaId)=>{
        // console.log(tareaId);
        const accion = {
            type:'delete',
            payload: tareaId
        };
        dispatch(accion);
    }

    const handleToggle = (tareaId)=>{
        // las tareas tienen una prop√¨edad llamada done, que sirve para saber si la tarea ya fue realizada o no, con ese metodo cambio su valor booleano
        dispatch({type:'toggle',payload: tareaId});
    }

    return (
        <div>
            <h1>Todo App ({tareas.length}) </h1>
            <hr/>
            <div className='row'>
                <div className='col-7'>
                <TareaLista tareas={tareas} handleDelete={handleDelete} handleToggle={handleToggle} />
                </div>
                <div className='col-5'>
                    <AgregarTarea handleAgregarTarea={handleAgregarTarea} />
                </div>
            </div>
        </div>
    )
}
