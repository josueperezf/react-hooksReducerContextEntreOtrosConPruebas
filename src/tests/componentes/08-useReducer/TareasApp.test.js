import { act } from "@testing-library/react";
import { mount, shallow } from "enzyme";
import { TareasApp } from "../../../componentes/08-useReducer/TareasApp";
import { listadoTareasDePrueba } from "../../fixtures/listadoTareasDePrueba";

describe('pruebas en <TareasApp/>',()=>{

    const wrapper = shallow(<TareasApp/>);
    // lo siguiente es para probar que realmente almacena en localstorage
    Storage.prototype.setItem = jest.fn(()=>{});

    test('Debe Mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('debe de agregar una tarea', () => {
        // shall es generar para probar un componente
        // mount es para probar toda la aplicacion en contexto, digamos que es para probar los componentes principales
        // mount da mas informacion sobre componentes hijos
        const wrapper = mount(<TareasApp/>);
        // agregando (listadoTareasDePrueba[0]); digamos que ejecuto el resultado de una vez
        //  a = wrapper.find('AgregarTarea').prop('handleAgregarTarea');
        // a(listadoTareasDePrueba[0]);
        act(()=>{
            wrapper.find('AgregarTarea').prop('handleAgregarTarea')(listadoTareasDePrueba[0]);
            wrapper.find('AgregarTarea').prop('handleAgregarTarea')(listadoTareasDePrueba[1]);
        });
        // Todo App ({tareas.length}) 
        expect(wrapper.find('h1').text().trim()).toBe(`Todo App (2)`);
        // probando si almaceno en el localshorage dos veces
        expect(localStorage.setItem).toHaveBeenCalledTimes(2)
    });

    test('Debe de eliminar una tarea', () => {
        // agregamos una tarea
        wrapper.find('AgregarTarea').prop('handleAgregarTarea')(listadoTareasDePrueba[0]);
        
        expect(wrapper.find('h1').text().trim()).toBe(`Todo App (1)`);
        // eliminamos una tarea, no podemos hacer esta prueba sin una tarea agregada, por ello se hace
        wrapper.find('TareaLista').prop('handleDelete')(listadoTareasDePrueba[0].id);
        expect(wrapper.find('h1').text().trim()).toBe(`Todo App (0)`);
    });
    
    
    
});