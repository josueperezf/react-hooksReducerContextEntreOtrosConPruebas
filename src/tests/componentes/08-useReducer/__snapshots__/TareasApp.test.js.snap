// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en <TareasApp/> Debe Mostrarse correctamente 1`] = `
<div>
  <h1>
    Todo App (
    0
    ) 
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TareaLista
        handleDelete={[Function]}
        handleToggle={[Function]}
        tareas={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <AgregarTarea
        handleAgregarTarea={[Function]}
      />
    </div>
  </div>
</div>
`;
